### ISSUE TITLE
Rethinking the design

### ISSUE THREAD

**Issue by maurolepore** [2DegreesInvesting/r2dii.plot.static#118]
By now I have spent quite some time working with this package and
I believe I understand its structure much better than before. It's only
now that I'm able to see the big picture and consider the overall user 
experience.

Here I reflect on the value of this package and use that insight to
redesign the API. Here I explain my idea; it open to discussion and
should be A-B tested against the current implementation and other
candidates. 

I'm pretty sure this will cause quite some stress but it should not as 
most changes I suggest could be implemented as thin
wrappers on top of what we already have.

@MonikaFu @AlexAxthelm, please comments with your thoughts or
let me know if you prefer to setup a call.

--

The most popular package to create plots is `ggplot2`.
`r2dii.plot.static` extends `ggplot2` to helps users of r2dii packages
create plots that are (a) insightful, (b) beautiful, and (c) quick and
easy to make. I think most of the value of this package lays on three
things:

1.  The variables we chose to plot, which most strongly determines
    insights.
2.  The theme, which most strongly determines beauty.
3.  The transformations we make to the user’s data, which most strongly
    determines how quickly a user can move from thought to plot.

Other aspects, such as adding specific labels, add little value because
many users would already know how to do it with `ggplot2`, and if they
don’t, they will benefit more from learning how to do it in `ggplot2` as
they can reuse that knowledge in ggplot2 and its extensions. Forcing
users to learn our new interfaces has a cognitive load that I would
rather users spend in more useful tasks.

### Proposed design

With this in mind I suggest this package should only provide the
following:

1.  Example datasets directly in data/. This way the datasets are
    already available when users use this package. This resembles the
    experience users should have when they use `r2dii.match` and
    `r2dii.analysis` to process their data. The output would be already
    available in their global environment and could be piped into
    plotting functions in this package.

2.  Themes, constructed as close as possible to standard themes in
    `ggplot2`. This way users can use 2DII themes with this package or
    wherever they use `ggplot2`.

3.  Quick plotting functions that produce a plot with a single argument
    (`data`), then extended with standard `ggplot2` functions. This way
    users can get 80% of what they want with 20% of the effort – i.e. a
    single function call. The remaining 20% of what they want they can
    get with plain `ggplot2` functions; that bit is very personal and we
    are unlikely to guess what it is. For this functions to be quick
    they’ll need to massage the data internally, expecting data with a
    very specific structure, failing fast if this requirement is
    violated, and hinting users about how they should fix the problem.

This API should deliver the most value per line of code. Adding more
stuff will likely result in a large code base that is hard to maintain
and an opinionated API that will be expensive for users to learn and
rigid compare to using `ggplot2` directly.

### Proposed README

The goal of `r2dii.plot` is to help you create insightful, beautiful,
and flexible plots as easily as possible. You can then extend them using
the most popular plotting package: `ggplot2`.

#### Installation

Install `r2dii.plot` from CRAN with:

``` r
install.packages("r2dii.plot")
```

#### Example (in pseudocode)

`r2dii.plot` extends the popular `ggplot2` package. You’ll commonly use both of them.

```r 
# (The package for interactive plots might be named r2dii.iplot) 
library(ggplot2)
library(r2dii.plot)
```

-   `theme_2dii()` helps you to beautify your existing ggplots. It
    implements 2DII style and the knowledge of 2DII’s visualization
    expert.

``` r
mtcars %>% 
  ggplot(aes(cyl)) +
  geom_histogram() +
  theme_2dii()
```

-   `plot_xyz()` helps you create an xyz plot from xyz-like data.

``` r
# r2dii.plot includes several example datasets
data <- r2dii_example("xyz")

data %>% 
  # All `plot_*() functions produce useful results with just one call
  plot_xyz() +
  # Extend your plot using the popular ggplot2 package directly
  labs(title = "Example of an xyz plot")
```

To see all functions visit <https://2DegreesInvesting.github.io/r2dii.plot>.

**Comment by MonikaFu** [comment]
In general I get your idea and support it, however many of the parameters that I ask users to set are more complex than simply adding `+ labs("label")` because they are often connected to the colors/categories present in the data and the order of the bars/scenarios etc. I would have to experiment to easy how easy these are to change on a graph that was created already. As in: can you do it without basically re-doing the entire chart.

**Comment by AlexAxthelm** [comment]
TL;DR: If the API for this package isn't entirely standard, I'm okay with that.

I don't think this package can be thought of in the same way as most common `ggplot2` extensions, because it it's not serving a common purpose. The extensions here only work in very specific contexts.

I think a good peer for comparison (chosen because it was the first good candidate alphabetically on CRAN) might be [ComplexUpset](https://cran.r-project.org/web/packages/ComplexUpset/vignettes/Examples_R.html). It's not a general purpose package, it exists to create a very specific type of plot, and as such, applying different themes is difficult, and the themes that it implements may not be useful when used outside of the context of this package. Similarly, most of the functions, plot layers and other objects that are defined in that package are only useful in that context.

If we discover later, that people want to use this package in conjunction with other ggplot extensions, and something is causing issues, we can prioritize those issues alongside everything else. 

I'm inclined to close with no further action.

**Comment by MonikaFu** [comment]
Thanks @AlexAxthelm. That was reassuring to see the example of that package. I think that we should keep at leats a part of that non-standard API. 

I'd say that it still might be useful for me and @maurolepore to look critically at the way I constructed the API and try to make it as intuitive and easy to use possible. Maybe there are some improvements as I didn't have user experience in mind very strongly when I was writing this code.

**Comment by maurolepore** [comment]
> I'm inclined to close with no further action. -- Alex

:+1:  Closing now

> it still might be useful for me and @maurolepore to look critically at the ... API -- Monika

Sounds good. I suggest we setup a call and follow from there.

**Comment by maurolepore** [comment]
Monika and I followed up this discussion on a call. Here are some reflections for the record:


1. For simple tasks, let users reuse existing knowledge of
   tidyverse packages. For example, users should be able to
   run `dplyr::filter(data, sector == "power")` to filter
   the power sector before plotting, and to run 
   `p + ggplot2::xlab("My x label")` to add a label to the 
   x axis.
   
1. Remove as many arguments as possible, e.g. arguments
   dealing with simple labels.

1. Simplify arguments as much as possible:

    - provide defaults when we can safely and sensibly guess, 
    e.g. we can guess the sector from the data and apply
    a sensible colour pallette.

    - prefer simpler inputs, e.g. prefer atomic vectors, over
    vectors, over lists/data frames.

1. Default conservatiely rather than destructively. For 
   example, if the user provides data for multiple sectors,
   default to plotting all sectors -- each in its own facet --
   rather than picking one sector and dropping the others.
